transport/src/main/java/io/netty/channel/SubChannel.java

```diff
@@ -0,0 +1,520 @@
+		return pipeline().write(msg, promise);
+	}
+
+	@Override
+	public ChannelFuture writeAndFlush(Object msg, ChannelPromise promise) {
+		return pipeline().writeAndFlush(msg, promise);
+	}
+
+	@Override
+	public ChannelFuture writeAndFlush(Object msg) {
+		return pipeline().writeAndFlush(msg);
+	}
+
+	@Override
+	public ChannelPromise newPromise() {
+		return pipeline().newPromise();
+	}
+
+	@Override
+	public ChannelProgressivePromise newProgressivePromise() {
+		return pipeline().newProgressivePromise();
+	}
+
+	@Override
+	public ChannelFuture newSucceededFuture() {
+		return pipeline().newSucceededFuture();
+	}
+
+	@Override
+	public ChannelFuture newFailedFuture(Throwable cause) {
+		return pipeline().newFailedFuture(cause);
+	}
+
+	@Override
+	public ChannelPromise voidPromise() {
+		return pipeline().voidPromise();
+	}
+
+	@Override
+	public int hashCode() {
+		return id().hashCode();
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		return this == o;
+	}
+
+	@Override
+	public int compareTo(Channel o) {
+		if (this == o) {
+			return 0;
+		}
+
+		return id().compareTo(o.id());
+	}
+
+	@Override
+	public String toString() {
+		return parent().toString() + "(subchannel)";
+	}
+
+	private final class SubChannelUnsafe implements Unsafe {
+		private final VoidChannelPromise unsafeVoidPromise = new VoidChannelPromise(SubChannel.this, false);
+		private boolean closeInitiated;
+
+		@Override
+		public void connect(final SocketAddress remoteAddress, SocketAddress localAddress,
+				final ChannelPromise promise) {
+			ctx.connect(remoteAddress, localAddress).addListener(new PromiseRelay(promise));
+		}
+
+		@Override
+		@SuppressWarnings("deprecation")
+		public Handle recvBufAllocHandle() {
+			return parent().unsafe().recvBufAllocHandle();
+		}
+
+		@Override
+		public SocketAddress localAddress() {
+			return parent().unsafe().localAddress();
+		}
+
+		@Override
+		public SocketAddress remoteAddress() {
+			return parent().unsafe().remoteAddress();
+		}
+
+		@Override
+		public void register(EventLoop eventLoop, ChannelPromise promise) {
+			if (!promise.setUncancellable()) {
+				return;
+			}
+			if (registered) {
+				throw new UnsupportedOperationException("Re-register is not supported");
+			}
+
+			registered = true;
+
+			promise.setSuccess();
+
+			pipeline().fireChannelRegistered();
+		}
+
+		@Override
+		public void bind(SocketAddress localAddress, ChannelPromise promise) {
+			ctx.bind(localAddress).addListener(new PromiseRelay(promise));
+		}
+
+		@Override
+		public void disconnect(ChannelPromise promise) {
+			ctx.disconnect().addListener(new PromiseRelay(promise));
+		}
+
+		@Override
+		public void close(final ChannelPromise promise) {
+			if (!promise.setUncancellable()) {
+				return;
+			}
+			if (closeInitiated) {
+				if (closeFuture().isDone()) {
+					// Closed already.
+					promise.setSuccess();
+				} else if (!(promise instanceof VoidChannelPromise)) { // Only needed if no VoidChannelPromise.
```

Comment:
How can reduce duplicate code here, and help keep this in sync with any future changes?